jdaviz/core/template_mixin.py

```diff
@@ -856,3 +832,163 @@ def __init__(self, *args, **kwargs):
+        super().__init__(plugin, label=label,
+                         label_default=label_default, label_auto=label_auto,
+                         label_invalid_msg=label_invalid_msg, label_overwrite=label_overwrite,
+                         add_to_viewer_items=add_to_viewer_items,
+                         add_to_viewer_selected=add_to_viewer_selected)
+
+        # DataCollectionAdd/Delete are fired even if remain unchecked in all viewers
+        self.hub.subscribe(self, DataCollectionAddMessage,
+                           handler=lambda _: self._on_label_changed())
+        self.hub.subscribe(self, DataCollectionDeleteMessage,
+                           handler=lambda _: self._on_label_changed())
+
+        self.viewer = ViewerSelect(plugin, add_to_viewer_items, add_to_viewer_selected,
+                                   manual_options=['None'])
+
+        self.add_observe(label, self._on_label_changed)
+        self.add_observe(label_default, self._on_label_default_changed)
+        self.add_observe(label_auto, self._on_label_auto_changed)
+
+    def _on_label_changed(self, msg={}):
+        if not len(self.label.strip()):
+            # strip will raise the same error for a label of all spaces
+            self.label_invalid_msg = 'label must be provided'
+            return
+
+        for data in self.app.data_collection:
+            if self.label == data.label:
+                if data.meta.get('Plugin', None) == self._plugin.__class__.__name__:
+                    self.label_invalid_msg = ''
+                    self.label_overwrite = True
+                    return
+                else:
+                    self.label_invalid_msg = 'label already in use'
+                    self.label_overwrite = False
+                    return
+
+        self.label_invalid_msg = ''
+        self.label_overwrite = False
+
+    def _on_label_default_changed(self, msg={}):
+        if self.label_auto:
+            self.label = self.label_default
+
+    def _on_label_auto_changed(self, msg={}):
+        if self.label_auto:
+            self.label = self.label_default
+
+    def add_results_from_plugin(self, data_item):
+        """
+        Add ``data_item`` to the app's data_collection according to the default or user-provided
+        label and adds to any requested viewers.
+        """
+        if self.label_invalid_msg:
+            raise ValueError(self.label_invalid_msg)
+        data_item.meta['Plugin'] = self._plugin.__class__.__name__
+
+        if self.label in self.app.data_collection:
+            self.app.data_collection.remove(self.app.data_collection[self.label])
+
+        self.app.add_data(data_item, self.label)
+
+        if self.add_to_viewer_selected != 'None':
+            # replace the contents in the selected viewer with the results from this plugin
+            # TODO: switch to an instance/classname check?
+            replace = self.viewer.selected_item.get('reference', None) != 'spectrum-viewer'
```

Comment:
I don't love `'spectrum-viewer'` being hardcoded here, but instance checks are difficult with avoiding circular imports.  Open to suggestions!